feat(styling): add theming details input feature (#846)
AmadeusITGroup/otter: docs/styling/THEME.md

>>>> **my-component-pres.style.scss**: component style - import style.theme.scss\
>>>> **my-component-pres.style.theme.scss**: variables - import app-styling.scss\
>>>> ...

## Technical structure (advance)

The function `apply-theme` expects a specific structure of [SCSS Map](https://sass-lang.com/documentation/values/maps/). To facilitate the generation of the latter, the function `meta-theme-to-otter` converts an [Angular Material Theme](https://material.angular.io/guide/theming) into a compatible structure.

The theme structure object should respect the following Schema:

```json
{
  "$ref": "#/definitions/varNode",

  "definitions": {

>     "varNode": {

Reviewer: to be honnest, it is a bit difficult to separate name that are just placeholders from actual expected properties

Author: I take all advice you can have on it :)

Reviewer: Maybe between "[ ]" ?
This is just an idea but something like 
"[yourVarNode]" : { ... ?

Author: So you suggest to not go with JSON Schema anymore?

Reviewer: yeah you are right, I guess it is not supported by the structure in general.

:root {
  @include o3r.var('o3r-primary-color', #000, (descriptions: 'My primary color used as website main color', label: 'My primary color', type: 'color', category: 'main color', tags: ('main', 'color', 'primary')));
  // As details parameter is optional, it can be reduced to `@include o3r.var('o3r-primary-color', #000)`
}
```

> **Note**: The mixin `o3r.var` is an alias of `o3r.define-var`.

Please beware that the **mixin** `o3r.var` and the **function** `o3r.var` are similar and made to work in different contexts:

```scss
@use '@o3r/styling' as o3r;

body {
  $myVariable: o3r.var('my-color', #000);
  background-color: $myVariable;
}
// will generate "background-color: var(--my-color, #000)" in CSS
// the purpose is to be able to use the variable --my-color without defining it value (but using the default value)
```

```scss
@use '@o3r/styling' as o3r;

body {
  @include o3r.var('my-color', #000);
  background-color: --my-color;
}
// will generate "--my-color: #000" in CSS
// the purpose is to be able to define the variable --my-color 
```

In both cases the variable will be extracted to the metadata with the specified default value.

### Override variable details

> The both function and mixins `o3r.var` can be used to override variable details.

Reviewer 2: ```suggestion
Both the function and the mixin `o3r.var` can be used to override variable details.
```

* **label**: Caption describing the variable, if not provided the CMS will use the variable name instead
* **type**: Additional context to help the CMS display the most relevant input widget (currently only `string` and `color` types are supported)
* **category**: Way to group different variables in the CMS
* **tags**: List of tags associated to the variable. It is a way of grouping variables or flagging them.

### Example

This is an example of a theme structure:

```scss
@use '@o3r/styling' as o3r;

$my-theme: (
  o3r: (
    primary: (
      color: (
        value: '#000',
        details: ( // Optional property
          descriptions: 'My primary color used as website main color', // Optional property
          label: 'My primary color', // Optional property
          type: 'color', // Optional property
          category: 'main color', // Optional property
          tags: ('main', 'color', 'primary') // Optional property
        )
      )
    )
  )
);

@include o3r.apply-theme($my-theme);
```

This will result in the following CSS:

```css
:root {
  --o3r-primary-color: #000;
}
```

### Define variable outside of theme mechanism

> In certain cases we want to define variable, make it part of the extracted metadata but without defining a full Otter Theme.

Reviewer: ```suggestion
In certain cases we want to define a variable and make it part of the extracted metadata without defining a full Otter Theme.
```

This can be achieved via the `o3r.var` mixin. If we take the previous example, the same result (in term of CSS and metadata) can be done as following:

```scss
@use '@o3r/styling' as o3r;

:root {

>   @include o3r.var('o3r-primary-color', #000, (descriptions: 'My primary color used as website main color', label: 'My primary color', type: 'color', category: 'main color', tags: ('main', 'color', 'primary')));

Reviewer: ```suggestion
  @include o3r.var('o3r-primary-color', #000, (description: 'My primary color used as website main color', label: 'My primary color', type: 'color', category: 'main color', tags: ('main', 'color', 'primary')));
```

}
```

> **Note**: The mixin `o3r.var` is an alias of `o3r.define-var`.

Please beware that the **mixin** `o3r.var` and the **function** `o3r.var` are similar and made to work in different contexts:

```scss
@use '@o3r/styling' as o3r;

body {
  $myVariable: o3r.var('my-color', #000);
  background-color: $myVariable;
}
// will generate "background-color: var(--my-color, #000)" in CSS
// the purpose is to be able to use the variable --my-color without defining it value (but using the default value)
```

```scss
@use '@o3r/styling' as o3r;

body {
  @include o3r.var('my-color', #000);
  background-color: o3r.var('my-color'); // equivalent to `--my-color`
}
// will generate "--my-color: #000" in CSS
// the purpose is to be able to define the variable --my-color
```

In both cases the variable will be extracted to the metadata with the specified default value.

### Override variable details

Both the function and the mixin `o3r.var` can be used to override variable details.

```scss
@use '@o3r/styling' as o3r;

// generate CSS variables from a theme:
@include o3r.apply-theme($my-theme);

:root {
  // override the value and details of the variable "my-variable" from the theme:

>   @include o3r.var('my-variable', #000, (descriptions: 'new description'));

Reviewer: ```suggestion
  @include o3r.var('my-variable', #000, (description: 'new description'));
```

  // override only the details of the variable "my-variable" from the theme:

>   @include o3r.var('my-variable', null, (descriptions: 'new description'));

Reviewer: ```suggestion
  @include o3r.var('my-variable', null, (description: 'new description'));
```

}

// override only the details of the variable "my-variable" from the theme:

> $res: o3r.var('my-variable', null, (descriptions: 'new description'));

Reviewer: ```suggestion
$res: o3r.var('my-variable', null, (description: 'new description'));
```

>>>> **my-component-pres.style.scss**: component style - import style.theme.scss\
>>>> **my-component-pres.style.theme.scss**: variables - import app-styling.scss\
>>>> ...

## Technical structure (advance)

> The function `apply-theme` is expected a specific structure of [SCSS Map](https://sass-lang.com/documentation/values/maps/). To facilitate its generations, the function `meta-theme-to-otter` is helping to convert an [Angular Material Theme](https://material.angular.io/guide/theming) to a compatible structure.

Reviewer: ```suggestion
The function `apply-theme` expects a specific structure of [SCSS Map](https://sass-lang.com/documentation/values/maps/). To facilitate the generation of said structure, the function `meta-theme-to-otter` converts an [Angular Material Theme](https://material.angular.io/guide/theming) into a compatible structure.
```

Reviewer 3: @cpaulve-1A of the latter ? 
I would simplify and say "To facilitate its generation" (close to the current text) or "To facilitate the generation of this structure"

> This is an example of a theme structure:

Reviewer: I feel that the structure is not so explicit for someone who doesn't know the framework theming. You just provide an example on what a theme structure can look like but I feel that will not help them so much.

Author: Added JSON schema to specify the expected structure

```scss
$my-theme: (
  o3r: (
    primary: (
      color: (
        value: '#000',
        details: ( // Optional property
          descriptions: 'My primary color used as website main color', // Optional property
          label: 'My primary color', // Optional property
          type: 'color', // Optional property
          category: 'main color', // Optional property
          tags: ('main', 'color', 'primary'), // Optional property
        )
      )
    )
  )
);

@include apply-theme($my-theme);
```

> This will result to the following CSS:

Reviewer 3: ```suggestion
This will result in the following CSS:
```

```css
:root {
  --o3r-primary-color: #000;
}
```

> The additional information, specified into the `details` property is used at extraction time to provide variable context information display in the CMS.

Reviewer 3: ```suggestion
The additional information, specified in the `details` property, is used at extraction time to provide variable context information to display in the CMS.
```

> The additional information, specified into the `details` property is used at extraction time to provide variable context information display in the CMS.

Reviewer: ```suggestion
The additional information, specified in the `details` property, is used at extraction time to provide context to the CMS regarding the theming variable.
```
@sdo-1A Would that be better?

Reviewer 3: ```suggestion
The additional information, specified in the `details` property, is used at extraction time to provide context regarding the theming variable to be displayed in the CMS.
```
I would word it like this

The **Tags** property contains the following information:

* **description**: Medium/long description of the variable
* **label**: Caption describing the variable, if not provided the CMS will use the variable name instead

> * **type**: indication of the variable to help customizing the CMS widget (currently only `string` an `color` types are supported)

Reviewer 3: ```suggestion
* **type**: indication of the variable to help CMS widget customization (currently only `string` and `color` types are supported)
```

> * **type**: indication of the variable to help customizing the CMS widget (currently only `string` an `color` types are supported)

Reviewer: ```suggestion
* **type**: Additional context to help the CMS display the most relevant input widget (currently only `string` an `color` types are supported)
```

> * **type**: indication of the variable to help CMS widget customization (currently only `string` and `color` types are supported)

Reviewer: ```suggestion
* **type**: Additional context to help the CMS display the most relevant input widget (currently only `string` an `color` types are supported)
```

> * **category**: Way to group different variable to assign them category

Reviewer 3: ```suggestion
* **category**: Way to group different variables to assign them a category
```

> * **category**: Way to group different variable to assign them category

Reviewer: ```suggestion
* **category**: Way to group different variables to assign them category
```

> * **category**: Way to group different variables to assign them category

Reviewer: ```suggestion
* **category**: Way to group different variables in the CMS
```

> * **tags**: List of tags associated to the variable. It is a way regroup variables of to flag them.

Reviewer 3: ```suggestion
* **tags**: List of tags associated to the variable. It is a way of grouping variables or flagging them.
```
